Filesystem [1/N]

Ghabry opened the pull request:
This is a partial patchset from the VFS changes rebased to upstream.
It only changes how File input/output works across Player in a consistent way by now. You don't get any new features by this but it makes future VFS work muuuuuch easier when these stream changes are kept being updated upstream.

For reviewing I recommend to look at the change as a whole. The changes by ChrisBreiti are a bit old :)

- All uses of ``FILE*`` replaced with C++ streams
- All C++ streams use now ``Filesystem::InputStream/Filesystem::OutputStream`` which is by now a shallow wrapper around ``std::filebuf``. 

Basicly the VFS code will do the following later:

1. You open a file like ``/home/user/game.zip/RPG_RT.ldb``
2. This returns a ``Filesystem::InputStream`` that inherits from ``istream``. So most of C++ fstream API will work as expected.
3. Use normal istream read funcs on it.

Is a draft. Need to test if I broke any Read/Write code. All these raw buffer APIs are hard to get right when a shared pointer is mixed in ^^'. Type confusion.

Ghabry commented:
Jenkins: Test this please

Ghabry commented:
There are still some crashes to resolve e.g. when project/save path are invalid.

But besides this: rdy for review

fmatthew5876 commented on src/bitmap.cpp:
```diff
@@ -106,21 +106,21 @@ Bitmap::Bitmap(const std::string& filename, bool transparent, uint32_t flags) {
 		Output::Warning("Unsupported image file %s (Magic: %02X)", filename.c_str(), *reinterpret_cast<uint32_t*>(data));
 
-	fclose(stream);
+	stream.reset(); //Deletes the stream and closes the file
```
Not strictly needed, unless you have a reason to want to close the file early and release the resources.

fmatthew5876 commented on src/filefinder.cpp:
```diff
@@ -668,6 +668,32 @@ std::shared_ptr<std::fstream> FileFinder::openUTF8(const std::string& name,
+std::shared_ptr<std::ostream> FileFinder::openUTF8Output(const std::string& name,
+	std::ios_base::openmode m)
+{
+	std::shared_ptr<std::ofstream> ret(new std::ofstream(
```
`std::make_shared`

fmatthew5876 commented on src/filefinder.cpp:
```diff
@@ -668,6 +668,32 @@ std::shared_ptr<std::fstream> FileFinder::openUTF8(const std::string& name,
+std::shared_ptr<std::istream> FileFinder::openUTF8Input(const std::string& name,
+	std::ios_base::openmode m)
+{
+	std::shared_ptr<std::ifstream> ret(new std::ifstream(
```
`std::make_shared`

fmatthew5876 commented on src/image_png.cpp:
```diff
@@ -31,6 +31,13 @@ static void read_data(png_structp png_ptr, png_bytep data, png_size_t length) {
 }
 
+static void read_data_istream(png_structp png_ptr, png_bytep data, png_size_t length) {
+	std::istream* bufp = (std::istream*)png_get_io_ptr(png_ptr);
```
Use `auto` on the left and `reintrerpret_cast` on the right.

fmatthew5876 commented on src/image_png.cpp:
```diff
@@ -31,6 +31,13 @@ static void read_data(png_structp png_ptr, png_bytep data, png_size_t length) {
 
+static void read_data_istream(png_structp png_ptr, png_bytep data, png_size_t length) {
+	std::istream* bufp = (std::istream*)png_get_io_ptr(png_ptr);
+	if (bufp != NULL&&*bufp) {
```
`nullptr` and whitespace

fmatthew5876 commented on src/output.cpp:
```diff
@@ -249,8 +249,8 @@ bool Output::TakeScreenshot() {
 bool Output::TakeScreenshot(std::string const& file) {
-	std::shared_ptr<std::fstream> ret =
-		FileFinder::openUTF8(file, std::ios_base::binary | std::ios_base::out | std::ios_base::trunc);
+	std::shared_ptr<std::ostream> ret =
```
I would use `auto` here

fmatthew5876 commented on src/player.cpp:
```diff
@@ -1034,21 +1043,16 @@ std::string Player::GetEncoding() {
-			encodings = ReaderUtil::DetectEncodings(is);
-		}
+		auto ldb_stream = FileFinder::openUTF8Input(ldb, std::ios::ios_base::in | std::ios::ios_base::binary);
+		std::vector<std::string> encodings = ReaderUtil::DetectEncodings(*ldb_stream);
```
`auto`

fmatthew5876 commented on src/audio_decoder.cpp:
```diff
@@ -105,19 +104,18 @@ class WMAUnsupportedFormatDecoder : public AudioDecoder {
 const char wma_magic[] = { (char)0x30, (char)0x26, (char)0xB2, (char)0x75 };
 
-std::unique_ptr<AudioDecoder> AudioDecoder::Create(FILE* file, const std::string& filename, bool resample) {
+std::unique_ptr<AudioDecoder> AudioDecoder::Create(std::shared_ptr<FileFinder::istream> stream, const std::string& filename, bool resample) {
```
Is `AudioDecoder` taking ownership of the stream? If so I would just pass `istream&` and forget the `shared_ptr`.

fmatthew5876 commented:
So one general comment instead of a sea of nits.

I don't think we need to use `shared_ptr` for this, in fact we don't event need `unique_ptr`.

The `istream` interface as of C++11 is movable.

https://en.cppreference.com/w/cpp/io/basic_iostream/basic_iostream

Also istream/ostream itself already encapsulates the encapsulates the real I/O inside of filebuf. So using pointers is adding another level of indirection we may not need.

So what I would suggest you explore is if we can just directly return `FileFinder::istream` by value, utilizing it's move constructor. Then it behaves pretty much like a `unique_ptr` and we'll get slightly better performance through fewer indirections.

If it turns out we can't do this, I would try `unique_ptr` over shared ptr. I'm not sure we really need or even want shared semantics for file handles.

fmatthew5876 commented on src/filefinder.cpp:
```diff
@@ -646,26 +646,37 @@ void FileFinder::Quit() {
+	std::streamsize size = FileFinder::GetFileSize(name);
+	auto *buf = new std::filebuf();
+
+	Filesystem::InputStream ret(new Filesystem::InputStreamRaw(buf->open(
```
`std::make_shared`

fmatthew5876 commented on src/filesystem.h:
```diff
@@ -0,0 +1,88 @@
+		case SEEK_END:
+			return std::ios_base::end;
+		default:
+			assert(false);
```
This should fail to compile in debug builds. Sadly, you can't put `assert()` in constexpr functions.

Ghabry replied:
C++14 allows non-constexpr functions in constexpr when they are in an untaken codepath when a constant is passed in.

(Is described somewhere here according to other websites but no idea where: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3652.html )

fmatthew5876 replied:
As we discussed on irc, this is now working correctly in C++14.

fmatthew5876 commented on src/image_xyz.cpp:
```diff
@@ -70,13 +71,11 @@ bool ImageXYZ::ReadXYZ(const uint8_t* data, unsigned len, bool transparent,
+	long size = stream->get_size();
 	std::vector<uint8_t> buffer(size);
-	long size_read = fread((void*) &buffer.front(), 1, size, stream);
+	long size_read = stream->read(reinterpret_cast<char*>(&buffer.front()), size).gcount();
```
`buffer.data()`

fmatthew5876 commented on src/input.cpp:
```diff
@@ -57,10 +57,10 @@ bool Input::IsWaitingInput() { return wait_input; }
+		ButtonMappingArray buttons,
+		DirectionMappingArray directions,
+		const std::string& replay_from_path,
+		const std::string& record_to_path
```
whitespace change intentional?

Ghabry replied:
Rebase error :/

fmatthew5876 commented on src/output.cpp:
```diff
@@ -206,30 +206,28 @@ static void HandleErrorOutput(const std::string& err) {
-		if (in.tellg() > log_size) {
-			in.seekg(-log_size, std::ios_base::end);
+	auto in = FileFinder::OpenInputStream(FileFinder::MakePath(Main_Data::GetSavePath(), OUTPUT_FILENAME).c_str(),std::ios::ios_base::in);
+	if (in&&!in->bad()) {
```
whitespace

fmatthew5876 commented on src/output.cpp:
```diff
@@ -206,30 +206,28 @@ static void HandleErrorOutput(const std::string& err) {
+
+			auto out = FileFinder::OpenOutputStream(FileFinder::MakePath(Main_Data::GetSavePath(), OUTPUT_FILENAME).c_str(), std::ios::ios_base::out);
+			out->write(buf, read);
+			out.reset();
```
Don't need the calls to `reset`

fmatthew5876 commented on src/platform/3ds/3ds_audio.cpp:
```diff
@@ -67,15 +67,15 @@ void CtrAudio::BGM_Play(std::string const& file, int volume, int pitch, int fade
-	bgm_decoder = AudioDecoder::Create(filehandle, file);
-	if (bgm_decoder && bgm_decoder->Open(filehandle)) {
+	bgm_decoder = AudioDecoder::Create(filestream, file);
+	if (bgm_decoder && bgm_decoder->Open(filestream)) {
```
For the `bgm_decoder` why do we have to pass the stream to the create method and then open it again?

Ghabry replied:
strange API design: Create will return an instance of a suitable audio decoder and Open() will do the actual processing. Should be improved but not here (and not before this PR, I don't want to rebase this :P)

fmatthew5876 commented on src/audio_decoder.cpp:
```diff
@@ -105,19 +104,21 @@ class WMAUnsupportedFormatDecoder : public AudioDecoder {
+std::unique_ptr<AudioDecoder> AudioDecoder::Create(Filesystem::InputStream stream, const std::string& filename, bool resample) {
 	char magic[4] = { 0 };
-	if (fread(magic, 4, 1, file) != 1)
+	if (stream->read(magic, sizeof(magic)).gcount() == 0) {
```
should it be `!= sizeof(magic)` what if I tried to read a 1, 2 or 3 byte file?

Ghabry replied:
added now such a ReadIntoObj, just not using it alot

fmatthew5876 commented on src/filesystem.h:
```diff
@@ -0,0 +1,88 @@
+
+	private:
+		std::streamsize size;
+	};
```
As we saw in https://github.com/EasyRPG/liblcf/pull/343  `seek` and `tell` result in a system call every single time they are used. For I/O which does a lot of telling, if you track the file offset yourself you get a massive speedup. For I/O which does a lot of small seeking, a call to `read()` into an empty buffer can be faster than a small seek.

If we wanted to fix this at the stream level, it would have to be done by implementing our own custom replacement for `filebuf`. Something to consider for future PR's. Maybe it could speed up the audio system?

Once all this work is done, an optimized `filebuf` could be worth exploring.

Ghabry replied:
I think this will be useful to implement. Other streams like compressed archive streams that must be baked by memory for performance will have a fast tell so "Real" Filestreams should have the same improvements
Also even zip reading has the fstream as underlying stream so everything will profit from it.

fmatthew5876 replied:
For the file only implementation, one might be able to get away with simply creating a new streambuf which wraps `filebuf` and intercepts it's seek and tell calls. This would need to be serparately unit tested and benchmarked carefully. I would definatly save any attempt for later PR.

For the zip file version, the details are notfully clear to me. But the overall design you'd want separate stream objects for each use like the file case. Each of those streams can reference a shared backend with 1 file handle. The buffer could be an LRU or ARC cache shared by all streams, or each stream could have it's own buffer. For this use case having a `shared_ptr` hidden inside the stream class could make a lot of sense. The zipfile resources go away exactly when all streams are closed.

Or would it just be easier to open same zipfile multiple times as read-only? One could cache and index to the file somewhere so new streams can jump directly to the part of the file they need to read without scanning.

fmatthew5876 commented on src/filesystem.h:
```diff
@@ -0,0 +1,88 @@
+
+	class vfs_istream : public std::istream {
+	public:
+		explicit vfs_istream(std::streambuf* sb, std::streamsize size);
```
Implement or delete move operations. `istream` is movable, so I believe your class will inherit it's move constructor, and then if a stream is moved into a new one, you'll get double delete on your rdbuf.

fmatthew5876 commented on src/filefinder.cpp:
```diff
@@ -646,26 +646,37 @@ void FileFinder::Quit() {
+Filesystem::InputStream FileFinder::OpenInputStream(const std::string& name,
+	std::ios_base::openmode m)
+{
+	std::streamsize size = FileFinder::GetFileSize(name);
```
You should be able to get the file size from the `filebuf`, this can save you from having to open the file twice.

Ghabry replied:
I'm a bit indifferent here. Which of the ways if faster should be also explored via benchmarking. 
At least for compressed streams seek-to-end and back to begin is bad because it forces a full decompression that could be avoided in some cases.

Also the file size isn't really used a lot except for some buffer preallocation optimisations. Imo we can remove it completely later and just read until eof (but needs more testing if this breaks some libs)

fmatthew5876 commented on src/decoder_wildmidi.cpp:
```diff
@@ -272,7 +297,12 @@ WildMidiDecoder::WildMidiDecoder(const std::string file_name) {
 	}
 	Output::Debug("WildMidi: Using %s as configuration file...", config_file.c_str());
 
+#if LIBWILDMIDI_VERSION >= 1027 // at least 0.4.3
```
Do we need to support old wildmidi still?

Ghabry replied:
This feature will be only useful when you want to load instrument files from a non-filesystem path (e.g. zip or APK) and it was actually submitted by CBreiti for us.
So there is no reason to reject older wildmidi Versions here as they will still work perfectly fine.

fmatthew5876 commented on src/audio_sdl_mixer.cpp:
```diff
@@ -44,6 +44,58 @@
+	SDL_RWops* create_StreamRWOps(Filesystem::InputStream stream){
+		SDL_RWops * ret = SDL_AllocRW();
+		// create a new shared pointer to avoid deletion of the content when the scope of this function ends
+		ret->hidden.unknown.data1 = new RefHolder { std::move(stream) };
```
If this was `unique_ptr`, you could drop an indirection by releasing the `unique_ptr` here and deleting the stream itself in `vio_close`. No need for `RefHolder`

fmatthew5876 commented on src/bitmap.cpp:
```diff
@@ -116,11 +116,11 @@ Bitmap::Bitmap(const std::string& filename, bool transparent, uint32_t flags) {
 		Output::Warning("Unsupported image file %s (Magic: %02X)", filename.c_str(), *reinterpret_cast<uint32_t*>(data));
 
-	fclose(stream);
+	stream.reset();
```
No need for `reset` unless you want to close the file early on purpose.

fmatthew5876 commented on src/decoder_fmmidi.cpp:
```diff
@@ -50,15 +49,11 @@ int read_func(void* instance) {
-bool FmMidiDecoder::Open(FILE* file) {
-	this->file = file;
-
+bool FmMidiDecoder::Open(Filesystem::InputStream stream) {
```
If the decoder doesn't take ownership, it should pass by reference to avoid a refcount.

`unique_ptr` makes this easier as you'll get compiler errors unless you either pass it by reference for `std::move` it.

fmatthew5876 commented on src/decoder_opus.cpp:
```diff
@@ -24,26 +24,33 @@
-	return fclose(f);
+static opus_int64 vio_tell_func(void* stream) {
+	auto* f = reinterpret_cast<Filesystem::InputStreamRaw*>(stream);
+	return f->tellg();
```
I'm pretty sure this is `int64_t` everywhere, but maybe we want  a `static_cast<opus_int64>` just in case?

fmatthew5876 commented on src/decoder_wav.cpp:
```diff
@@ -30,24 +30,21 @@ WavDecoder::WavDecoder()
+	stream->seekg(6, std::ios::ios_base::cur);
 	uint16_t bitspersample;
-	fread(&bitspersample, 1, 2, file_);
+	stream->read(reinterpret_cast<char*>(&bitspersample), sizeof(bitspersample));
```
This is a good case where doing a single read into a buffer and copying the pieces out would be faster than all these individual reads and seeks.

fmatthew5876 commented:
So one general comment about the overall approach. I really like what you're doing here and I see the goal is to encapsulate I/O so we can load resources from anything. This is great.

We should carefully consider whether we really wanted shared ownership semantics here (i.e. `shared_ptr`).

For resources like files, you really want to know when they are opened and when they get closed. When you have a `shared_ptr`, this becomes difficult because anything could be holding a reference. It's possible to write a bug where you keep the file around too long. We should consider using `unique_ptr` instead. Not only will this make the resource lifetime more explicit, it'll also give us better performance as we don't pay for atomic ref counts anymore.

In fact we don't event need `unique_ptr`.

The `istream` interface as of C++11 is movable.

https://en.cppreference.com/w/cpp/io/basic_iostream/basic_iostream
https://en.cppreference.com/w/cpp/io/basic_istream/operator%3D

Also istream/ostream itself already encapsulates the real I/O inside of filebuf. So using pointers is adding another level of indirection we don't actually need.

Essentially, the `istream` itself can act as a `unique_ptr` for the `filebuf` it contains. By doing this you remove one level of indirection and we don't have to deal with pointers and null checks in our code.

Consider getting rid of the pointers and `RawInputStream` and just using `InputStream` directly by implementing the move operations and `std::move`ing the stream object itself when you want to pass ownership to another entity.

Ghabry commented:
Thanks for the long and detailed review appreciate it.
Because this is a huge and invasive change it is important to get the overall API correct in the first attempt to avoid more gigantic refactoring.

The usage of ``shared_ptr`` was recommended in a stackoverflow answer but your explanation makes sense. Shared ptr is not really needed here as there is always exclusive ownership on the stream. Will try to get everything (std)moved.

fmatthew5876 commented:
Another consideration.

I think at the end of this, we'll want the stream implementation in liblcf. Then we can build tools around zipping and unzip games etc..

If that part gets merged to player first, are we going to have to deal with license discussions to port it to liblcf?

Ghabry commented:
This sounds again as liblcf as the "catch all" library for everything that could be useful.

At least an efficient fstream would make sense in liblcf because this directly influences the parsing speed. (the Read(std::string) version of the API could use it, then all tools profit from it).

For zip: This adds dependencies to liblcf and zipping/unzipping games is out of scope for this PR. I will only do on-the-fly decompress in memory. Is not clear to me why they should use our unpack tool instead of unar?
The only packing use-case I see is somthing like "Make gamedisk" but this is an editor feature.

---
For everything that is related to "core" functionality like Video/Audio/Input/IO you can assume that I agree on MIT-licensing.

fmatthew5876 commented:
I think best way is to just proceed putting it in player and we can move pieces later as needed.

Ghabry commented:
Syntax question:
I do ``auto is = FileFinder::OpenInputStream (return InputStream&&)``. How do I move this now up the call hierarchy to make it "owned" by a unique_ptr?

Lets say I do "Audio::Open(is)". What kind of argument must Open accept? ``&`` or ``&&``. How do I assign it to a unique_ptr? I tried different combination with and without std::move and they all give compiler errors.

fmatthew5876 commented:
> Syntax question:
> I do `auto is = FileFinder::OpenInputStream (return InputStream&&)`. How do I move this now up the call hierarchy to make it "owned" by a unique_ptr?

Not sure what you mean here. That syntax is not valid C++.

> 
> Lets say I do "Audio::Open(is)". What kind of argument must Open accept? `&` or `&&`. How do I assign it to a unique_ptr? I tried different combination with and without std::move and they all give compiler errors.

If `Open()` is going to take ownership of the stream, have it take the `unique_ptr<InputStream>` by value. At the callsite, you move it in. If it's going to just use the stream, take it as either `InputStream*` or `InputStream&`.

Ghabry commented:
still WIP, just addressed the shared_ptr part. Also crashes on startup in LOG_FILE for some reason.

I added this shared_ptr -> Movable as a new commit because this is terrible to rebase. If you don't want this I have to squash everything upto "Add Filesystem::Input/OutputStream and use it everywhere."

fmatthew5876 commented on src/input_source.cpp:
```diff
@@ -88,7 +88,7 @@ bool Input::Source::InitRecording(const std::string& record_to_path) {
 		auto path = record_to_path.c_str();
 
-		record_log = FileFinder::OpenOutputStream(path, std::ios::out | std::ios::trunc);
+		record_log = std::make_unique<Filesystem_Stream::OutputStream>(FileFinder::OpenOutputStream(path, std::ios::out | std::ios::trunc));
```
This doesn't make sense, you create an `OutputStream` temporary, and then move it into a unique_ptr.

Now that your `OutputStream` itself is movable we don't need the `unique_ptr` at all.

Just do 

`auto record_log = FileFinder::OpenOutputStream(path, std::ios::out | std::ios::trunc);`

If someone wants to use the stream:
````
void use(OutputStream& s);
use(record_log)
```

If someone want so take ownership
```
take(OutputStream s)
take(std::move(s));
```

fmatthew5876 commented on src/filesystem_stream.h:
```diff
@@ -0,0 +1,170 @@
+		}
+		InputStream& operator=(InputStream&& is) {
+			std::istream::operator=(std::move(is));
+			set_rdbuf(is.rdbuf());
```
Need to test against self assignment here. i.e. `x = std::move(x)`

fmatthew5876 commented on src/filesystem_stream.h:
```diff
@@ -0,0 +1,170 @@
+		InputStream& operator=(const InputStream&) = delete;
+		InputStream(InputStream&& is) : std::istream(std::move(is)) {
+			set_rdbuf(is.rdbuf());
+			is.set_rdbuf(nullptr);
```
You forgot to move `streamsize`. Since we're setting rdbuf to nullptr in `is`, we should also make `is.streamsize = 0`

fmatthew5876 commented on src/filesystem_stream.h:
```diff
@@ -0,0 +1,170 @@
+#include "utils.h"
+
+namespace Filesystem_Stream {
+	class InputStream : public std::istream {
```
Since these stream objects are going to be used direct and not behind a pointer, should we make them `final` so people don't try to inherit from them and get slicing behavior?

Also if this is final. the `size` member can be `private` and not `protected`

Ghabry replied:
The ``protected`` was a logic error. I thought the Filesystem-Implementations will inherit from it but they will all only implement there own buffer objects, so this can be made ``final+private``

Ghabry commented:
Imo this is ready for reviewing & careful playtesting (any IO errors) on different platforms.

I will still improve some stuff but this is 99% ready.

fmatthew5876 commented:
I tested some games with this including HH3. LGTM!

This one is a potential source of rebase hell. So I would prefer it goes in sooner rather than later.

Ghabry commented:
I try to remove the GetSize() call. In all except two places it can be replaced with ``Utils::ReadStream``.

Problematic is libsndfile and SDL RW Ops. They both want a filesize function. But I guess both will be happy without a size information because not all streams are seekable or have a known size.

fmatthew5876 commented:
Optimization which can be done later:

- [ ] `Utils::ReadStream` should take the vector by reference and fill it in. This will allow clients to reuse a single buffer and not allocate every time. You can provide 2 overloads, an optimal one which the user passes the buffer, and a convenience once which allocates and returns it.
- [ ] For large files, growing linearly is not efficient. Try starting with 8k and doubling the size each time.

Ghabry commented:
The XMP IsModule test does not go through the Filesystem stuff.
I submitted a patch to them in January 2019 which is not merged yet :(
https://github.com/cmatsuoka/libxmp/pull/134

Ghabry commented:
XMP is finally being developed again and my patch was integrated :+1:
